File: Oficinas/29-07-22 Introdução à Programação Competitiva/Códigos/Diamantes e Areia/diamantes.cpp
#include<bits/stdc++.h>

using namespace std;

int main(){
    int n, i;
    string s;

    cin >> n;
    while(n--){
        cin >> s;

        stack<char> pilha;  //pilha para auxiliar na tarefa deste problema
        int resp = 0;
        //Percorrendo todos os caracteres de nossa string
        for(i = 0; i < s.size(); i++){
            if (s[i] == '<')        //Se o caractere for um <
                pilha.push(s[i]);   //Empilhamos ele
            else if ((s[i] == '>') && !pilha.empty()){
                //Se o caractere for > e a pilha não estiver vazia,
                //é pq temos um "diamante", como <....>
                pilha.pop(); //tiramos o < da pilha
                resp++;      //e incrementamos a resposta
            }
        }
        cout << resp << "\n";
    }
}

File: Oficinas/29-07-22 Introdução à Programação Competitiva/Códigos/Diamantes e Areia/diamantes.py
from collections import deque #módulo Python com a implementação de um deque (double queue)
                              #que também pode ser usado como uma pilha ou uma fila

n = int(input())
for k in range(n):
    s = input()
    pilha = deque() #deque para auxiliar na tarefa deste problema (utilizando como pilha)
    resp = 0
    #Percorrendo todos os caracteres de nossa string
    for i in range(0, len(s)):
        if s[i] == '<':         #Se o caractere for um <
            pilha.append(s[i])  #empilhamos ele
        elif s[i] == '>' and len(pilha) > 0:
            #Se o caractere for > e a pilha não estiver vazia,
            #é pq temos um "diamante", como <....>
            pilha.pop()     #tiramos o < da pilha
            resp = resp + 1 #e incrementamos a resposta
    print(resp)